### PR TITLE
Enable `torch.ops.aten.slice_scatter.default`  lowering

### PR DESCRIPTION
**Pull request by pschang-phy**
### Ticket
https://github.com/tenstorrent/pytorch2.0_ttnn/issues/485

### Problem description
`torch.ops.aten.slice_scatter.default`  remains not lowering to TTNN.

### What's changed
`torch.ops.aten.slice_scatter.default(input, src, dim, start, end)` could be lowered into `ttnn.slice` + `ttnn.concat`
or
just return `src` tensor, if size of `src` tensor == size of `input` tensor

The only 2 models, **ghostnet_100.in1k** and **ghostnetv2_100.in1k** which has `torch.ops.aten.slice_scatter.default` op are the case of just return `src` tensor

Test
```bash
pytest -s tests/lowering/tensor_manipulation/test_slice_scatter.py
```

### CODE REVIEW

File: torch_ttnn/passes/lowering/to_tt_pass.py [hunk 1]
```diff
@@ -1048,6 +1048,49 @@ def batch_norm_inference(input, weight, bias, mean, var, momentum, eps):
...
+                tensor_shape = tensor.meta["val"].size()
+                src_tensor_shape = src_tensor.meta["val"].size()
+
+                # nothing to slice, result tensors == src_tensor
+                if tensor_shape == src_tensor_shape:
+                    return src_tensor
+
+                # slice_scatter could be concat([pre_slice_tensor, src_tensor, post_slice_tensor])
+                rank = len(tensor_shape)
+                end = min(tensor_shape[dim], end) if end >= 0 else (end + tensor_shape[dim])
```
Comment (jerrysky3) [inline torch_ttnn/passes/lowering/to_tt_pass.py:1063]: The `end` might still be out of range if it is too small on the negative side?

```suggestion
                end = end if end >= 0 else (end + tensor_shape[dim])
                end = 0 if end < 0 else min(tensor_shape[dim], end)
```

File: torch_ttnn/passes/lowering/to_tt_pass.py [hunk 2]
```diff
@@ -1048,6 +1048,49 @@ def batch_norm_inference(input, weight, bias, mean, var, momentum, eps):
...
+                [step] = step or [1]
+
+                if step != 1 or dim >= rank:
```
Comment (jerrysky3) [inline torch_ttnn/passes/lowering/to_tt_pass.py:1066]: Is `dim >= rank` supported by `torch.slice_scatter`? If not maybe we should use assert to fail fast here?

File: torch_ttnn/passes/lowering/to_tt_pass.py [hunk 3]
```diff
@@ -1048,6 +1048,49 @@ def batch_norm_inference(input, weight, bias, mean, var, momentum, eps):
...
+                    return None
+
+                tensors_list = []
+                # pre_slice_tensor
+                if start > 0:
+                    slice_start = np.zeros(rank, dtype=int)
```
Comment (jerrysky3) [inline torch_ttnn/passes/lowering/to_tt_pass.py:1072]: IMO `[0] * rank` is a little bit simpler. I'm fine with both ways

File: torch_ttnn/passes/lowering/to_tt_pass.py [hunk 4]
```diff
@@ -1048,6 +1048,49 @@ def batch_norm_inference(input, weight, bias, mean, var, momentum, eps):
...
+                    slice_end = list(tensor_shape)
+                    slice_end[dim] = start
+                    tensors_list.append(g.call_function(ttnn.slice, (tensor, [*slice_start], [*slice_end])))
+
+                # src_tensor
+                tensors_list.append(src_tensor)
+
+                # post_slice_tensor
+                if end < tensor_shape[dim]:
+                    slice_start = np.zeros(rank, dtype=int)
```
Comment (jerrysky3) [inline torch_ttnn/passes/lowering/to_tt_pass.py:1082]: Same here

File: torch_ttnn/passes/lowering/to_tt_pass.py [hunk 5]
```diff
@@ -1048,6 +1048,49 @@ def batch_norm_inference(input, weight, bias, mean, var, momentum, eps):
...
+                if step != 1 or dim >= rank:
+                    return None
+
+                tensors_list = []
+                # pre_slice_tensor
+                if start > 0:
+                    slice_start = np.zeros(rank, dtype=int)
+                    slice_end = list(tensor_shape)
+                    slice_end[dim] = start
+                    tensors_list.append(g.call_function(ttnn.slice, (tensor, [*slice_start], [*slice_end])))
```
Comment (jerrysky3) [inline torch_ttnn/passes/lowering/to_tt_pass.py:1075]: If both `slice_start` and `slice_end` are list, we can directly pass them?

```suggestion
                    tensors_list.append(g.call_function(ttnn.slice, (tensor, slice_start, slice_end)))
```

File: torch_ttnn/passes/lowering/to_tt_pass.py [hunk 6]
```diff
@@ -1048,6 +1048,49 @@ def batch_norm_inference(input, weight, bias, mean, var, momentum, eps):
...
+
+                # src_tensor
+                tensors_list.append(src_tensor)
+
+                # post_slice_tensor
+                if end < tensor_shape[dim]:
+                    slice_start = np.zeros(rank, dtype=int)
+                    slice_start[dim] = end
+                    slice_end = list(tensor_shape)
+                    tensors_list.append(g.call_function(ttnn.slice, (tensor, [*slice_start], [*slice_end])))
```
Comment (jerrysky3) [inline torch_ttnn/passes/lowering/to_tt_pass.py:1085]: Same here

**Review by jerrysky3 — APPROVED** [review]

File: torch_ttnn/passes/lowering/to_tt_pass.py [hunk 7]
```diff
@@ -1048,6 +1048,49 @@ def batch_norm_inference(input, weight, bias, mean, var, momentum, eps):
                 output_shape = node.meta["val"].size()
                 return insert_sharded_nhwc_to_nchw(g, output_tensor, output_shape)
 
+            if node.target == torch.ops.aten.slice_scatter.default:
+                tensor, src_tensor, dim, start, end, *step = args
```
Comment (swimdi) [inline torch_ttnn/passes/lowering/to_tt_pass.py:1052]: [start & end seems optional in there](https://pytorch.org/docs/stable/torch.compiler_ir.html), and maybe appeared in kwargs, maybe should have a check

File: torch_ttnn/passes/lowering/to_tt_pass.py (hunk 2 above)
Comment (pschang-phy) [reply torch_ttnn/passes/lowering/to_tt_pass.py]: ok, I have changed to assert back early.

File: torch_ttnn/passes/lowering/to_tt_pass.py (hunk 3 above)
Comment (pschang-phy) [reply torch_ttnn/passes/lowering/to_tt_pass.py]: changed to `[0] * rank`, it's more simpler

File: torch_ttnn/passes/lowering/to_tt_pass.py (hunk 5 above)
Comment (pschang-phy) [reply torch_ttnn/passes/lowering/to_tt_pass.py]: I have revised it.

File: torch_ttnn/passes/lowering/to_tt_pass.py (hunk 7 above)
Comment (pschang-phy) [reply torch_ttnn/passes/lowering/to_tt_pass.py:1052]: I have revised to more check on `start` and `end`

**Pull request merged**